### PR TITLE
ECAL SC parameters to EventSetup

### PR DESCRIPTION
**Pull request by thomreis**
#### PR description:

This PR moves the up to now hard coded ECAL mustache SC parameters and dynamic dPhi parameters to the EvetnSetup so that they can be taken from the GT in the future. A separate condFormat is defined for both sets of parameters and two helper classes are used to access the data within the SC code. Since the new records are not yet in the DB they are provided with an ESSource and ESProducer.
Data maker modules to generate sqlite files are included.
The initial parameters in this PR are the same as the ones currently in the master branch, introduced with #31823 , and so no changes are expected in the output.

#### PR validation:

- All runTheMatrix.py -l limited -i all --ibeos tests passed
- Perfect agreement between SCs reconstructed with CMSSW_11_2_0_pre7 with #31823 parameters and this code on top of CMSSW_11_2_0_pre7 (RelValH125GG sample, 2500 events). [SC energy](https://treis.web.cern.ch/treis/cms/ecal_dpg/mustache_params/histos_single_2500ev/spectrum_superCluster_energy_xMax500.png) (green: reference, red: this PR, blue: different parameter set)

### CODE REVIEW

**Comment by slava77** [comment]
> Since the new records are not yet in the DB they are provided with an ESSource and ESProducer.
>  Data maker modules to generate sqlite files are included.

why not use a DBsource with a toGet pointing to an sqlfile?

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by thomreis** [comment]
> > Since the new records are not yet in the DB they are provided with an ESSource and ESProducer.
> > Data maker modules to generate sqlite files are included.
> 
> why not use a DBsource with a toGet pointing to an sqlfile?

Making the record on the fly at this stage avoids having to deal with sqlite files when in the future the record will be in the GT anyway. The ESProducer is also used to produce sqlite files if needed. Furthermore, being able to change the parameters in the configuration of the ESProducer will make retuning easier.

**Comment by jpata** [comment]
@cmsbuild please test

File: RecoEcal/EgammaCoreTools/interface/Mustache.h [hunk 1]
```diff
@@ -5,18 +5,30 @@
...
+#include "RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h"
+#include "RecoEcal/EgammaCoreTools/interface/SCDynamicDPhiParametersHelper.h"
 
 namespace reco {
   namespace MustacheKernel {
-    bool inMustache(
-        const float maxEta, const float maxPhi, const float ClustE, const float ClusEta, const float ClusPhi);
-    bool inDynamicDPhiWindow(
-        const float seedEta, const float seedPhi, const float ClustE, const float ClusEta, const float clusPhi);
+    bool inMustache(const std::shared_ptr<MustacheSCParametersHelper>& params,
```
Comment (jpata) [inline RecoEcal/EgammaCoreTools/interface/Mustache.h:13]: could you pass `const MustacheSCParametersHelper*` instead, to avoid making a copy of the object in `updateSCParams`?

File: RecoEcal/EgammaCoreTools/interface/Mustache.h [hunk 2]
```diff
@@ -5,18 +5,30 @@
...
+                    const float maxEta,
+                    const float maxPhi,
+                    const float ClustE,
+                    const float ClusEta,
+                    const float ClusPhi);
+    bool inDynamicDPhiWindow(const std::shared_ptr<SCDynamicDPhiParametersHelper>& params,
```
Comment (jpata) [inline RecoEcal/EgammaCoreTools/interface/Mustache.h:19]: could you pass `const SCDynamicDPhiParametersHelper*` instead, to avoid making a copy of the object in `updateSCParams`?

**Review by jpata — COMMENTED** [review]
I was wondering about the use of shared_ptr for storing the parameters in the PF algos. It looks to me the accessor methods of `EcalMustacheSCParameters` and `SCDynamicDPhiParametersHelper` are thread-safe, so you wouldn't need to make copies (as I think the current approach is doing) in beginRun. Possibly I misunderstand the intent here, let me know.

File: RecoEcal/EgammaCoreTools/src/Mustache.cc [hunk 3]
```diff
@@ -33,114 +23,67 @@ namespace reco {
...
+      const float a_upper = (1 / (4 * curv_up)) - std::abs(b_upper);
+      const float a_lower = (1 / (4 * curv_low)) - std::abs(b_lower);
 
       const double dphi = TVector2::Phi_mpi_pi(ClusPhi - maxPhi);
       const double dphi2 = dphi * dphi;
       // minimum offset is half a crystal width in either direction
       // because science.
-      const float upper_cut = (std::max((1. / (4. * a_upper)), 0.0) * dphi2 + std::max(b_upper, 0.0087f)) + 0.0087;
-      const float lower_cut = (std::max((1. / (4. * a_lower)), 0.0) * dphi2 + std::min(b_lower, -0.0087f));
+      constexpr float half_crystal_width = 0.0087;
```
Comment (jpata) [inline RecoEcal/EgammaCoreTools/src/Mustache.cc:52]: In addition, I think a comment or reference to the origin would be needed here. Is it the same parameter as https://github.com/cms-sw/cmssw/blob/9c9e72f134d6727f96fbbed3a538b4c408e2e8d6/L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc#L79

File: RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#ifndef RecoEcal_EgammaCoreTools_MustacheSCParametersHelper_h
+#define RecoEcal_EgammaCoreTools_MustacheSCParametersHelper_h
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "CondFormats/EcalObjects/interface/EcalMustacheSCParameters.h"
+
+namespace reco {
+
+  class MustacheSCParametersHelper : public EcalMustacheSCParameters {
```
Comment (makortel) [inline RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h:9]: I find this pattern somewhat confusing. To be more precise
* the `static_cast<EcalMustacheSCParameters>()` in the ESProducer looks weird (even if it is technically correct)
* the creation of `MustacheSCParametersHelper` from `EcalMustacheSCParameters` leads to copying all the data

Would it be feasible to move `sqrtLogClustETuning()` and `parabolaParameters()` to `EcalMustacheSCParameters` directly?

If not, how about making `MustacheSCParametersHelper` to hold `EcalMustacheSCParameters` by (`const`) pointer to avoid the copying? (the helper could even be produced by and ESProducer to avoid repeating its construction in different modules and EDM streams).

File: RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                                                                                            float absSeedEta) const {
+  // assume the collection is sorted in descending ParabolaParameters.etaMin and descending ParabolaParameters.minEt
+  for (const auto &parabolaParams : parabolaParametersCollection_) {
+    if (log10ClustE < parabolaParams.log10EMin || absSeedEta < parabolaParams.etaMin) {
+      continue;
+    } else {
+      return parabolaParams;
+    }
+  }
+  return EcalMustacheSCParameters::ParabolaParameters();
```
Comment (makortel) [inline RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc:48]: How about returning `const EcalMustacheSCParameters::ParabolaParameters*` to avoid copying many vectors? (or `std::optional<std::reference_wrapper<EcalMustacheSCParameters::ParabolaParameters>>`)

File: RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+void MustacheSCParametersHelper::setSqrtLogClustETuning(const float sqrtLogClustETuning) {
+  sqrtLogClustETuning_ = sqrtLogClustETuning;
+}
+
+EcalMustacheSCParameters::ParabolaParameters MustacheSCParametersHelper::parabolaParameters(float log10ClustE,
+                                                                                            float absSeedEta) const {
+  // assume the collection is sorted in descending ParabolaParameters.etaMin and descending ParabolaParameters.minEt
+  for (const auto &parabolaParams : parabolaParametersCollection_) {
+    if (log10ClustE < parabolaParams.log10EMin || absSeedEta < parabolaParams.etaMin) {
```
Comment (makortel) [inline RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc:42]: How many elements `parabolaParametersCollection_` is expected to have? (just thinking if binary search would be worth it)

Comment (thomreis) [reply RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc:42]: I was wondering if there is a more efficient way of doing this. We have a set parameters in the pipeline (not in this PR) that would have 15*15 parameters. I think this will be fine grained enough for the future.

File: RecoEcal/EgammaCoreTools/src/Mustache.cc (hunk 3 above)
Comment (thomreis) [reply RecoEcal/EgammaCoreTools/src/Mustache.cc:52]: To be honest I do not know. It certainly looks very similar but I do not think the origin is in the L1Trigger package. The value comes from the detector geometry and I thought the variable name is descriptive enough.

File: CondFormats/EcalObjects/src/classes.h [hunk 7]
```diff
@@ -221,5 +223,13 @@ namespace CondFormats_EcalObjects {
...
 
     EcalSimPulseShape ecal_sim_pulse_shapes;
+
+    EcalMustacheSCParameters ecal_mustache_sc_parameters;
+    EcalMustacheSCParameters::ParabolaParameters ecal_mustache_sc_parabola_parameters;
+    std::vector<EcalMustacheSCParameters::ParabolaParameters> v_ecal_mustache_sc_parabola_parameters;
+
+    EcalSCDynamicDPhiParameters ecal_sc_dynamic_dphi_parameters;
+    EcalSCDynamicDPhiParameters::DynamicDPhiParameters ecal_sc_dynamic_dphi_parameters_struct;
+    std::vector<EcalSCDynamicDPhiParameters::DynamicDPhiParameters> v_ecal_sc_dynamic_dphi_parameters;
```
[the remark this answers is not in the document]
Comment (thomreis) [reply CondFormats/EcalObjects/src/classes.h:233]: You mean no need to change this file?

File: RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc (hunk 6 above)
Comment (makortel) [reply RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc:42]: Thanks. 225 elements sounds like a binary search with `std::lower_bound` should be faster on average.

File: CondFormats/EcalObjects/src/classes.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (makortel) [reply CondFormats/EcalObjects/src/classes.h:233]: Or you could remove the `namespace CondFormats_EcalObjects { ... }` block.

[the remark this answers is not in the document]
Comment (thomreis) [reply CondFormats/EcalObjects/src/classes.h:233]: Maybe a bit out of the scope of this PR but I can do it since I am anyway going to change the condFormats/EcalObjects package.

File: RecoEcal/EgammaCoreTools/interface/Mustache.h (hunk 1 above)
Comment (thomreis) [reply RecoEcal/EgammaCoreTools/interface/Mustache.h:13]: Yes, I think that is possible.
You mean getting a pointer to pass with
`mustacheSCParamsHelper_ = static_cast<const reco::MustacheSCParametersHelper*>(ecalMustacheSCParamsHandle_.product());`?

File: RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h (hunk 4 above)
Comment (thomreis) [reply RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h:9]: The intention was that the `EcalMustacheSCParameters` in CondFormats` holds the data and all manipulations of the data are done with the `MustacheSCParametersHelper` in the `RecoEcal` package.

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc [hunk 8]
```diff
@@ -145,6 +154,15 @@ void PFECALSuperClusterAlgo::update(const edm::EventSetup& setup) {
...
 }
 
+void PFECALSuperClusterAlgo::updateSCParams(const edm::EventSetup& setup) {
+  edm::ESHandle<EcalMustacheSCParameters> ecalMustacheSCParamsHandle_ = setup.getHandle(ecalMustacheSCParametersToken_);
+  mustacheSCParamsHelper_ = std::make_shared<reco::MustacheSCParametersHelper>(*ecalMustacheSCParamsHandle_.product());
+  edm::ESHandle<EcalSCDynamicDPhiParameters> ecalSCDynamicDPhiParamsHandle_ =
+      setup.getHandle(ecalSCDynamicDPhiParametersToken_);
+  scDynamicDPhiParamsHelper_ =
+      std::make_shared<reco::SCDynamicDPhiParametersHelper>(*ecalSCDynamicDPhiParamsHandle_.product());
+}
```
[the remark this answers is not in the document]
Comment (thomreis) [reply RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:164]: This could be a good solution because these parameters will probably only change in the order of once a year.

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc [hunk 9]
```diff
@@ -145,6 +154,16 @@ void PFECALSuperClusterAlgo::update(const edm::EventSetup& setup) {
   channelStatus_ = esChannelStatusHandle_.product();
 }
 
+void PFECALSuperClusterAlgo::updateSCParams(const edm::EventSetup& setup) {
+  edm::ESHandle<EcalMustacheSCParameters> ecalMustacheSCParamsHandle_ = setup.getHandle(ecalMustacheSCParametersToken_);
+  mustacheSCParamsHelper_ =
+      static_cast<const reco::MustacheSCParametersHelper*>(ecalMustacheSCParamsHandle_.product());
```
Comment (makortel) [inline RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:160]: The ESProduct is `EcalMustacheSCParameters`, it can not be casted to a derived class type.

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (thomreis) [reply RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:161]: Is using an `ESWatcher` still the preferred method to update the payload now that no helper is involved anymore in retrieving the parameters?

**Comment by jpata** [comment]
@cmsbuild please test

File: RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc (hunk 6 above)
Comment (thomreis) [reply RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc]: How would the search have to be set up in this case? `std::lower_bound` by default gives back an iterator to the element where `value <= element`. In particular if value > the largest element in the collection the iterator points to end(). What I would need is `element <= value` such that e.g. with two sets of parameters binned in eta (set 1 has etaMin = 0 and set 2 has etaMin = 1.5) for a value of eta = 2.1 I get back the second set and for eta = 0.3 I get back the first set.

I tried the following but that leads to the above mentioned problem.
```
  const auto vars = std::make_pair(log10ClustE, absSeedEta);                                                                                                  
  const auto it = std::lower_bound(parabolaParametersCollection_.begin(),                                                                                     
                                   parabolaParametersCollection_.end(),                                                                                       
                                   vars,                                                                                                                      
                                   [](const EcalMustacheSCParameters::ParabolaParameters &params,                                                             
                                      const std::pair<double, double> vars) {                                                                                 
                                     const auto bounds = std::make_pair(params.log10EMin, params.etaMin);                                                     
                                     return bounds > vars;                                                                                                    
                                   });                                                                                                                        
  return (it != parabolaParametersCollection_.end()) ? &*it : nullptr;
```

**Comment by thomreis** [comment]
Looks like the tests are stuck.

**Comment by christopheralanwest** [comment]
please test

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (makortel) [reply RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:161]: No, the `ESWatcher` is not useful for just getting products from the EventSetup.

File: RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc (hunk 6 above)
Comment (makortel) [reply RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc]: After playing a bit with a toy code I think I see what you mean, i.e. `lower_bound()` returns an `element <= value`, but when the energy of `value` is larger than largest `energy` of `element`s, the eta of `element` can be larger than eta of  `value`. Even in that case the `element` found by `lower_bound()` could be used for a starting point of a linear search to find the correct box.

I'm not sure if the added complexity is worth of the effort though, so the search probably best to leave as it is unless a profiler shows it to be expensive.

File: RecoEcal/EgammaClusterProducers/src/PFECALSuperClusterProducer.cc [hunk 10]
```diff
@@ -159,6 +159,10 @@ PFECALSuperClusterProducer::PFECALSuperClusterProducer(const edm::ParameterSet&
 
 PFECALSuperClusterProducer::~PFECALSuperClusterProducer() {}
 
+void PFECALSuperClusterProducer::beginRun(Run const& run, EventSetup const& iE) {
+  superClusterAlgo_.updateSCParams(iE);
```
Comment (makortel) [inline RecoEcal/EgammaClusterProducers/src/PFECALSuperClusterProducer.cc:163]: If the EventSetup objects are used only in the `produce()` function, the recommended pattern is to get the products only in the `produce()` function. Then the `ESGetToken`s won't need `edm::Transition::BeginRun` either.

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc [hunk 11]
```diff
@@ -145,6 +152,14 @@ void PFECALSuperClusterAlgo::update(const edm::EventSetup& setup) {
   channelStatus_ = esChannelStatusHandle_.product();
 }
 
+void PFECALSuperClusterAlgo::updateSCParams(const edm::EventSetup& setup) {
+  edm::ESHandle<EcalMustacheSCParameters> ecalMustacheSCParamsHandle_ = setup.getHandle(ecalMustacheSCParametersToken_);
+  mustacheSCParams_ = ecalMustacheSCParamsHandle_.product();
```
Comment (makortel) [inline RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc:157]: These could be simplified to
```suggestion
  mustacheSCParams_ = &setup.getData(ecalMustacheSCParametersToken_);
```

File: RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc (hunk 6 above)
Comment (thomreis) [reply RecoEcal/EgammaCoreTools/src/MustacheSCParametersHelper.cc]: I was able to get it to work with a two step technique. The first search finds the right energy bin and the second search then looks for the matching eta bin. For the second search the range is already reduced since one does not need to look beyond the matched energy bins.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by thomreis** [comment]
Hi @jpata can you stop the tests and launch them again in about 5 min please? I had to amend on little change. It should be OK now.

**Comment by jpata** [comment]
@cmsbuild please cancel

**Comment by jpata** [comment]
OK no problem, let me know when this is ready to be tested.

**Comment by thomreis** [comment]
@jpata now it should be ready to start the tests.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
> @cmsbuild please cancel

@jpata. the command is "please abort"

**Comment by jpata** [comment]
For the moment, all reco comparisons fail (do not show results) due to an issue with the IB. I think we need to wait ~O(24h) and submit again.

**Comment by jpata** [comment]
@cmsbuild please test

File: CondTools/Ecal/python/EcalMustacheSCParametersMaker_cfg.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (jpata) [inline CondTools/Ecal/python/EcalMustacheSCParametersMaker_cfg.py:1]: I think this file might belong better in /test/, it doesn't seem like something you would want to import from other configs.

File: CondTools/Ecal/python/EcalSCDynamicDPhiParametersMaker_cfg.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (jpata) [inline CondTools/Ecal/python/EcalSCDynamicDPhiParametersMaker_cfg.py:1]: I think this file might belong better in /test/, it doesn't seem like something you would want to import from other configs.

File: RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h (hunk 4 above)
Comment (jpata) [reply RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h:9]: @thomreis I think the principle you describe makes sense to me, however, I'm trying to understand if the implementation could perhaps be simplified further.

Is the only functionality of the helper to prepare the data `EcalMustacheSCParameters` from the config using `MustacheSCParametersHelper::MustacheSCParametersHelper(const edm::ParameterSet &iConfig)`?

If so, can you remove the helper class entirely and just use an utility function located in e.g. RecoEcal/EgammaCoreTools to prepare EcalMustacheSCParameters in the ESProducer?

File: RecoEcal/EgammaCoreTools/src/Mustache.cc (hunk 3 above)
Comment (jpata) [reply RecoEcal/EgammaCoreTools/src/Mustache.cc:55]: OK, that sounds reasonable to me. Anyway your implementation of making it a constexpr is already an improvement. Perhaps in the future this number might be accessed from geometry.

File: RecoEcal/EgammaCoreTools/src/Mustache.cc [hunk 14]
```diff
@@ -33,114 +26,70 @@ namespace reco {
...
+      const float upper_cut =
+          (std::max((1. / (4. * a_upper)), 0.0) * dphi2 + std::max(b_upper, half_crystal_width)) + half_crystal_width;
+      const float lower_cut = (std::max((1. / (4. * a_lower)), 0.0) * dphi2 + std::min(b_lower, -half_crystal_width));
 
       //if(deta < upper_cut && deta > lower_cut) inMust=true;
```
Comment (jpata) [inline RecoEcal/EgammaCoreTools/src/Mustache.cc:60]: might be removed

**Review by jpata — COMMENTED** [review]
Thank you for the updates to this PR! Tests look good! I had some further comments inline about the implementation.

File: CondTools/Ecal/python/EcalMustacheSCParametersMaker_cfg.py (hunk 12 above)
Comment (thomreis) [reply CondTools/Ecal/python/EcalMustacheSCParametersMaker_cfg.py:1]: It seemed to be the right place. There are many similar looking cfg.py files in the python directory. But I can move it to /test as well. Does not make a difference as long as we remember where to find it.

File: RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h (hunk 4 above)
Comment (thomreis) [reply RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h:9]: The `EcalMustacheSCParameters` has only getter functions but no setters at the moment. Initially, even the getters were in the helper but this was changed to avoid copying the objects.
This makes changes on the ECAL side possible (e.g. changes number of parameters in case we want to use a higher order polynomial at some point) without touching the code in the CondFormats directory.
Having a utility function to prepare the conditions from the config file would mean to include setters in `EcalMustacheSCParameters` which would make it possible to set the conditions directly without the checks and sorting that the helper does. Alternatively, the helper could be made a friend class in RecoEcal could contain the setters.

File: CondTools/Ecal/python/EcalMustacheSCParametersMaker_cfg.py (hunk 12 above)
Comment (jpata) [reply CondTools/Ecal/python/EcalMustacheSCParametersMaker_cfg.py:1]: My understanding (which is far from complete) has been that we put in /python/ general "building blocks" for configuration, whereas in /test/ various standalone runnable configs.

Doing a check in the reco packages at least (CondTools is not reco), I see that cms.Process is very rarely defined in the /python/ dir. Perhaps there is a different approach in CondTools or AlCa packages in general.

```
$ find . -wholename "./Reco*/*/python/*" -name "*.py" -exec grep -H "cms\.Process" {} \; | wc -l
5
$ find . -wholename "./Reco*/*/test/*" -name "*.py" -exec grep -H "cms\.Process" {} \; | wc -l
317
```

Comment (thomreis) [reply CondTools/Ecal/python/EcalMustacheSCParametersMaker_cfg.py:1]: I guess this has not been handled in the same way all the time. And different packages might handle things different on top of it. EDM modules are also not always in the plugins directory. Anyway, I have moved the cfg files to the test directory now where they are fine as well.

File: RecoEcal/EgammaCoreTools/src/Mustache.cc (hunk 3 above)
Comment (thomreis) [reply RecoEcal/EgammaCoreTools/src/Mustache.cc:55]: OK I will leave it this way then.

File: RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h (hunk 4 above)
Comment (makortel) [reply RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h:9]: To me a `friend` class would be clearer pattern than using inheritance for a deriving class to set members of the base class and then slice the base part of the object with `static_cast`.

A `friend` would be more restrictive, now anyone could fill the contents of the base class in arbitrary way by inheriting with a custom class (even if that could be unlikely for anyone to do in practice).

Comment (thomreis) [reply RecoEcal/EgammaCoreTools/interface/MustacheSCParametersHelper.h]: I have implemented the `friend` solution now.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by thomreis** [comment]
Comparison failures are unrelated to this PR as far as I can tell.

**Comment by jpata** [comment]
Yes, this logfile stuff is related to xrootd randomness and is irrelevant.

**Comment by jpata** [comment]
@thomreis we discussed in reco and we should try to avoid friend classes if at all possible, as it breaks encapsulation. In addition, I find the pattern where this constructor `MustacheSCParametersHelper(EcalMustacheSCParameters &params, const edm::ParameterSet &iConfig);` actually initializes `EcalMustacheSCParameters` somewhat surprising and hard to follow.

Here's what I had in mind with the utility function in this gist, let me know if this could work, or perhaps I missed something:
https://gist.github.com/jpata/a821e8d5640c0938828cd62ca890ad0b

**Comment by thomreis** [comment]
@jpata the helper as a friend allowed that the `EcalMustacheSCParameters` can only be set via the helper. The correct setting of the parameters therefore ensured. Introducing the `init` function (or maybe, alternatively, a dedicated constructor) allows `EcalMustacheSCParameters` with non-standard parameters to be created.

**Comment by jpata** [comment]
OK, fair point. So `MustacheSCParametersHelper` is a builder class that ensures that the immutable object `EcalMustacheSCParameters` is built correctly from only a single place. Some more questions to understand the use of this pattern here:

- What is the lifetime of `MustacheSCParametersHelper`, is it only to be used from `EcalMustacheSCParametersESProducer::EcalMustacheSCParametersESProducer`?

- Are there any concerns with passing an already-initialized `EcalMustacheSCParameters` to `MustacheSCParametersHelper::MustacheSCParametersHelper`?

- Why do you need the following constructor: `MustacheSCParametersHelper::MustacheSCParametersHelper(EcalMustacheSCParameters &params)`

**Comment by thomreis** [comment]
> OK, fair point. So `MustacheSCParametersHelper` is a builder class that ensures that the immutable object `EcalMustacheSCParameters` is built correctly from only a single place. Some more questions to understand the use of this pattern here:
> 
>     * What is the lifetime of `MustacheSCParametersHelper`, is it only to be used from `EcalMustacheSCParametersESProducer::EcalMustacheSCParametersESProducer`?
In my initial PR the helpers were also used to get the data from the ES but to avoid copying this has been moved to the condFormats directly. So in this PR only `EcalMustacheSCParametersESProducer` still uses `MustacheSCParametersHelper`. However, certain user codes use the mustache parameters function in a standalone way and for those there should be a method to build a valid `MustacheSCParameters` from a config file.
> 
>     * Are there any concerns with passing an already-initialized `EcalMustacheSCParameters` to `MustacheSCParametersHelper::MustacheSCParametersHelper`?
Good point. I need to clear the vector before adding the new parameters sets. Otherwise there will be duplicates.
> 
>     * Why do you need the following constructor: `MustacheSCParametersHelper::MustacheSCParametersHelper(EcalMustacheSCParameters &params)`
This is a relic from the initial PR. I think it can be removed.

**Comment by jpata** [comment]
> Good point. I need to clear the vector before adding the new parameters sets. Otherwise there will be duplicates.

> This is a relic from the initial PR. I think it can be removed.

Thanks for explanations, this is helpful! If you could address these points for both the EcalMustacheSCParameters and EcalSCDynamicDPhiParameters, I would appreciate it!

in addition, here https://github.com/cms-sw/cmssw/pull/32066/files#diff-12b338205d355f289cf103a5bd2bad2e68534a72eba2b34b845ea34d95a05258R20
you are redefining `params` with a different type, maybe better to call it `parabola_params` or such.

**Comment by jpata** [comment]
@cms-sw/alca-l2 @cms-sw/db-l2 could you take a look at this PR from your side, if possible? From the reco side, it's mostly finalized (code cleanup still todo), and we would like to make an effort to get this in by pre10 (due tomorrow).

**Comment by thomreis** [comment]
Ah no changing the base release added some commits I did not want here.

**Comment by kpedro88** [comment]
-1
@thomreis please follow http://cms-sw.github.io/tutorial-resolve-conflicts.html

**Comment by thomreis** [comment]
OK, this looks better. Sorry for the noise.

**Comment by christopheralanwest** [comment]
please test

**Comment by jpata** [comment]
+reconstruction

- migrates ECAL Mustache supercluster parameters to the EventSetup by creating new CondFormats for `EcalMustacheSCParameters` and `EcalSCDynamicDPhiParameters`
- the new conditions are provided currently by ESSource & ESProducer (not yet in global tag)
- tests pass with no differences, as expected

**Comment by christopheralanwest** [comment]
please rebase to fix conflicts

**Comment by jpata** [comment]
Looks like https://github.com/cms-sw/cmssw/pull/32253 got merged in the mean time, yes, @thomreis please fix & I will resign for reco.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by thomreis** [comment]
The failed comparisons are not related to changes in this PR.

**Comment by ggovi** [comment]
+1

**Comment by jpata** [comment]
+reconstruction
 
  - migrates ECAL Mustache supercluster parameters to the EventSetup by creating new CondFormats for EcalMustacheSCParameters and EcalSCDynamicDPhiParameters
  - the new conditions are provided currently by ESSource & ESProducer (not yet in global tag)
  - tests show DQM differences from unrelated items (other merged PRs in the test)

**Comment by thomreis** [comment]
Conflict after #31726 was merged has been resolved.
Can you launch the tests again please?

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
@silviodonato @qliphy can you make sure this gets in the pre10 release if possible? this would be necessary to make sure the GT can be prepared later AFAIK. the last two rebases and delays in the last 24h have been due to other PRs getting merged and creating conflicts.

**Comment by silviodonato** [comment]
urgent

**Comment by Martin-Grunewald** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by thomreis** [comment]
@christopheralanwest, @ggovi can you please sign again after the rebase?

**Comment by thomreis** [comment]
Crossed mid air. :-)

**Comment by jpata** [comment]
hmm, there's a lot of changes, but only in the workflow 11642.911, which seems to be for DD4HEP. However, I can't see any such PR in the bot report: https://github.com/cms-sw/cmssw/pull/32066#issuecomment-733809748. Any ideas?

**Comment by mmusich** [comment]
> Any ideas?

https://github.com/cms-sw/cmssw/issues/32283

**Comment by jpata** [comment]
ah, thank you, it was not on my radar.

**Comment by jpata** [comment]
+reconstruction

- migrates ECAL Mustache supercluster parameters to the EventSetup by creating new CondFormats for `EcalMustacheSCParameters` and `EcalSCDynamicDPhiParameters`
- the new conditions are provided currently by ESSource & ESProducer (not yet in global tag)
- tests show many differences, but they are unrelated/fake per #32283

**Comment by thomreis** [comment]
Links to plots in the comparison seem broken.

**Comment by christopheralanwest** [comment]
+1

**Comment by silviodonato** [comment]
+1
